benchmark/src/pre_benchmarks.cpp

```diff
@@ -0,0 +1,319 @@
+    }
+}
+
+void PRE_DecryptBeforeReEncrypt(benchmark::State &state) {
+    for (auto _ : state) {
+        cryptoContext->Decrypt(keyPairproducer.secretKey, ciphertext1, &plaintextDecproducer);
+    }
+    plaintextDecproducer->SetLength(plaintext->GetLength());
+}
+
+void PRE_Rekeygen(benchmark::State &state) {
+    keyPairconsumer = cryptoContext->KeyGen();
+
+    for (auto _ : state) {
+        benchmark::DoNotOptimize(reencryptionKey =
+            cryptoContext->ReKeyGen(keyPairproducer.secretKey, keyPairconsumer.publicKey));
+    }
+}
+
+void PRE_ReEncrypt(benchmark::State &state) {
+    for (auto _ : state) {
+        if (security_model == 0) {
+            //std::cout << "CPA secure PRE" << std::endl;
+            benchmark::DoNotOptimize(reEncryptedCT = cryptoContext->ReEncrypt(ciphertext1, reencryptionKey)); //IND-CPA secure
+        }
+        else if (security_model == 1) {
+            //std::cout << "Fixed noise (20 bits) practically secure PRE" << std::endl;
+            benchmark::DoNotOptimize(reEncryptedCT = cryptoContext->ReEncrypt(ciphertext1, reencryptionKey, keyPairproducer.publicKey)); //fixed bits noise HRA secure
+        } else if ((security_model == 2) || (security_model == 3)) {
+            //std::cout << "Provable HRA secure PRE" << std::endl;
+            benchmark::DoNotOptimize(reEncryptedCT1 = cryptoContext->ReEncrypt(ciphertext1, reencryptionKey, keyPairproducer.publicKey)); //HRA secure noiseflooding
+            benchmark::DoNotOptimize(reEncryptedCT = cryptoContext->ModReduce(reEncryptedCT1)); //mod reduction for noise flooding
+        } else {
+            std::cerr << "Not a valid security mode" << std::endl;
+            exit(EXIT_FAILURE);
+        }
+    }
+}
+
+void PRE_DecryptAfterReEncrypt(benchmark::State &state) {
+    for (auto _ : state) {
+        cryptoContext->Decrypt(keyPairconsumer.secretKey, reEncryptedCT, &plaintextDecconsumer);
+    }
+    plaintextDecconsumer->SetLength(plaintext->GetLength());
+}
+
+
+int main(int argc, char** argv)
+{
+   char opt(0);
+    static struct option long_options[] =
+    {
+        {"Security model",       required_argument, NULL, 'm'},
+        {"Number of hops",       required_argument, NULL, 'd'},
+        {"help",                 no_argument,       NULL, 'h'},
+        {NULL, 0, NULL, 0}
+    };
+
+    const char* optstring = "m:d:h";
+    while ((opt = getopt_long(argc, argv, optstring, long_options, NULL)) != -1) {
+        std::cerr << "opt1: " << opt << "; optarg: " << optarg << std::endl;
+        switch (opt) {
+        case 'm':
+            security_model = atoi(optarg);
+            break;
+        case 'd':
+            num_of_hops = atoi(optarg);
+            break;
+        case 'h':
+            usage();
+        default:
+            return false;
+        }
+    }
+
+    //0 - CPA secure PRE, 1 - fixed 20 bits noise, 2 - provable secure HRA noise flooding
+    parameters.SetPREMode(INDCPA);
```

Comment:
parameters.SetPREMode(INDCPA);
parameters.SetKeySwitchTechnique(BV); are set before the if statements because they would be the default. If I set them together in one place after the if statement, then it would replace the option set within the if statement for the specific conditions.

Reply:
set your parameters once, please don't re-set them